File: clients/lumify-sdk-python/lumify/sse.py
"""
Server-Sent Events reader for ``GET /v1/events/{id}/stream`` (see
api/routes/v1/stream.py). The endpoint emits named events — ``score`` on
change, ``done`` when the event finishes, ``error`` if the id doesn't exist —
plus unnamed ``: keep-alive`` comment pings every ~15s.

This parses the wire format directly off the HTTP response body (stdlib
``urllib``); it does not use any EventSource abstraction, so it can send the
``Authorization`` header the API requires. Mirrors the TS SDK's sse.ts.
"""

from __future__ import annotations

import codecs
import json
import socket
import urllib.error
import urllib.request
from typing import Any, Iterable, Iterator, Optional

from .errors import ConnectionError, error_from_response

_KEEPALIVE_FLOOR = 60.0  # keep the read timeout above the server's ~15s pings


class SSEEvent:
    """A single parsed SSE frame."""

    __slots__ = ("event", "data", "id")

    def __init__(self, event: str, data: str, id: Optional[str] = None) -> None:  # noqa: A002
        #: The ``event:`` field (defaults to "message" per the SSE spec).
        self.event = event
        #: Raw ``data:`` field, joined across multiline data.
        self.data = data
        self.id = id


class ScoreStreamEvent:
    """A decoded live-score stream event: ``event`` is one of
    ``"score"`` | ``"done"`` | ``"error"`` | ``"reconnect"``; ``data`` is the
    parsed JSON. ``"reconnect"`` is handled transparently by
    :func:`stream_scores` / :func:`lumify.aio.sse.astream_scores` (they
    reconnect and keep yielding) — it's surfaced here only for visibility."""

    __slots__ = ("event", "data")

    def __init__(self, event: str, data: Any) -> None:
        self.event = event
        self.data = data

    def __repr__(self) -> str:  # pragma: no cover - cosmetic
        return "ScoreStreamEvent(event=%r, data=%r)" % (self.event, self.data)


def _parse_frame(raw: str) -> Optional[SSEEvent]:
    data_lines = []
    event = "message"
    ident: Optional[str] = None
    comment_only = True

    for line in raw.split("\n"):
        if line.startswith(":"):
            continue  # comment / keep-alive
        comment_only = False
        if line.startswith("event:"):
            event = line[len("event:"):].strip()
        elif line.startswith("data:"):
            data_lines.append(line[len("data:"):].lstrip())
        elif line.startswith("id:"):
            ident = line[len("id:"):].strip()

    if comment_only and not data_lines:
        return None
    return SSEEvent(event=event, data="\n".join(data_lines), id=ident)


def parse_sse_stream(chunks: Iterable[bytes]) -> Iterator[SSEEvent]:
    """Parse a ``text/event-stream`` byte stream into individual SSE frames.

    Exposed for advanced use and testing; most callers want
    :func:`stream_scores`.
    """
    decoder = codecs.getincrementaldecoder("utf-8")()
    buffer = ""
    for chunk in chunks:
        if not chunk:
            continue
        buffer += decoder.decode(chunk)
        buffer = buffer.replace("\r\n", "\n").replace("\r", "\n")
        while "\n\n" in buffer:
            raw, buffer = buffer.split("\n\n", 1)
            frame = _parse_frame(raw)
            if frame is not None:
                yield frame

    tail = buffer + decoder.decode(b"", final=True)
    tail = tail.replace("\r\n", "\n").replace("\r", "\n").strip("\n")
    if tail:
        frame = _parse_frame(tail)
        if frame is not None:
            yield frame


def _iter_bytes(fp: Any, size: int = 1024) -> Iterator[bytes]:
    while True:
        chunk = fp.read(size)
        if not chunk:
            break
        yield chunk


def _connect(client: Any, event_id: int, connect_timeout: Optional[float]):
    url = client.build_url("/v1/events/%d/stream" % event_id)

    base = client.timeout if connect_timeout is None else connect_timeout
    if base == 0:
        read_timeout: Optional[float] = None
    else:
        read_timeout = max(float(base), _KEEPALIVE_FLOOR)

    request = urllib.request.Request(
        url, headers=client.auth_headers("text/event-stream"), method="GET"
    )
    try:
        return urllib.request.urlopen(request, timeout=read_timeout)
    except urllib.error.HTTPError as exc:
        body = exc.read().decode("utf-8", "replace")
        payload = None
        try:
            doc = json.loads(body)
            if isinstance(doc, dict) and isinstance(doc.get("error"), dict):
                payload = doc["error"]
        except ValueError:
            pass
        request_id = exc.headers.get("x-request-id") if exc.headers else None
        raise error_from_response(exc.code, payload, request_id)
    except (socket.timeout, TimeoutError) as exc:
        raise ConnectionError(
            "SSE connect to /v1/events/%d/stream timed out." % event_id, cause=exc
        )
    except urllib.error.URLError as exc:
        raise ConnectionError(
            "SSE connect to /v1/events/%d/stream failed: %s"
            % (event_id, getattr(exc, "reason", exc)),
            cause=exc,
        )


def stream_scores(
    client: Any,
    event_id: int,
    *,
    connect_timeout: Optional[float] = None,
) -> Iterator[ScoreStreamEvent]:
    """Iterate live score updates for an event over SSE. Emits only on change
    (plus a final ``done``), so it's far cheaper than polling the score endpoint.

    The server caps a single connection at 5 minutes and sends an
    ``event: reconnect`` frame just before closing it for that reason (as
    opposed to ``event: done``, which means the game actually finished). This
    reconnects automatically and keeps yielding, so a long game's stream looks
    continuous to the caller — the ``reconnect`` event is still yielded for
    visibility/telemetry, but no action is required in response to it.

    Note (sync/zero-dep tradeoff vs. the TS SDK): stdlib ``urllib`` applies one
    socket timeout to the whole request, so the read timeout is held above the
    server's keep-alive interval rather than being connect-only. Pass
    ``connect_timeout=0`` to disable the timeout entirely.
    """
    while True:
        resp = _connect(client, event_id, connect_timeout)
        reconnecting = False
        with resp:
            for frame in parse_sse_stream(_iter_bytes(resp)):
                if frame.event == "message" and not frame.data:
                    continue  # keep-alive with no named event
                try:
                    data = json.loads(frame.data) if frame.data else None
                except ValueError:
                    continue  # malformed frame — skip rather than raise mid-stream
                yield ScoreStreamEvent(event=frame.event, data=data)
                if frame.event == "reconnect":
                    reconnecting = True
                    break  # tear this connection down below, then loop to reconnect
                if frame.event in ("done", "error"):
                    return
        if not reconnecting:
            return  # stream ended without an explicit reconnect signal

File: clients/lumify-sdk/src/sse.ts
// Server-Sent Events reader for GET /v1/events/{id}/stream (see api/routes/v1/stream.py).
// The endpoint emits named events — `score` on change, `done` when the event
// finishes, `error` if the event id doesn't exist — plus unnamed `: keep-alive`
// comment pings every ~15s. This module has no dependency on the DOM
// `EventSource` API (which can't set an Authorization header); it parses the
// wire format directly off the fetch `Response` body stream, which works
// identically in Node 18+ and browsers.

import type { LumifyClient } from "./client.js";
import { ConnectionError, errorFromResponse } from "./errors.js";
import type { LumifyErrorPayload } from "./errors.js";
import type { ScoreResponse } from "./generated/models.js";

export interface SSEEvent {
  /** SSE `event:` field. Defaults to "message" per the SSE spec if the server omits it. */
  event: string;
  /** Raw (still-a-string) `data:` field, joined across multiline data. */
  data: string;
  id?: string;
}

/**
 * Parse a `text/event-stream` response body into individual SSE frames.
 * Exposed for advanced use; most callers want {@link streamScores}.
 *
 * Frames are separated by a blank line (`\n\n` or `\r\n\r\n`). The reader is
 * cancelled when the consumer breaks out of the async iterator so the upstream
 * fetch connection is released promptly.
 */
export async function* parseSSEStream(body: ReadableStream<Uint8Array>): AsyncGenerator<SSEEvent, void, void> {
  const reader = body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });

      // Normalize CRLF → LF so a single splitter handles both wire forms.
      buffer = buffer.replace(/\r\n/g, "\n").replace(/\r/g, "\n");

      let sepIndex: number;
      while ((sepIndex = buffer.indexOf("\n\n")) !== -1) {
        const rawFrame = buffer.slice(0, sepIndex);
        buffer = buffer.slice(sepIndex + 2);
        const frame = parseFrame(rawFrame);
        if (frame) yield frame;
      }
    }
  } finally {
    // Cancel the underlying stream so breaking out of `for await` tears down
    // the HTTP connection instead of waiting for the server's 5-minute cap.
    try {
      await reader.cancel();
    } catch {
      // already closed / cancelled
    }
    try {
      reader.releaseLock();
    } catch {
      // already released
    }
  }
}

function parseFrame(raw: string): SSEEvent | null {
  const dataLines: string[] = [];
  let event = "message";
  let id: string | undefined;
  let isCommentOnly = true;

  for (const line of raw.split("\n")) {
    if (line.startsWith(":")) continue; // comment / keep-alive
    isCommentOnly = false;
    if (line.startsWith("event:")) event = line.slice("event:".length).trim();
    else if (line.startsWith("data:")) dataLines.push(line.slice("data:".length).trimStart());
    else if (line.startsWith("id:")) id = line.slice("id:".length).trim();
  }

  if (isCommentOnly && dataLines.length === 0) return null; // pure keep-alive frame
  return { event, data: dataLines.join("\n"), id };
}

export interface StreamScoresOptions {
  /** Abort the stream from the caller's side (e.g. on a UI unmount). */
  signal?: AbortSignal;
  /**
   * Timeout for the *initial* HTTP connect/headers only (ms). The stream itself
   * is long-lived and is not subject to this timeout. Defaults to the client's
   * `timeoutMs` (30s). Pass `0` to disable.
   */
  connectTimeoutMs?: number;
}

export type ScoreStreamEvent =
  | { event: "score"; data: ScoreResponse }
  | { event: "done"; data: { event_id: number } }
  | { event: "error"; data: { error: { code: string; message: string } } }
  | { event: "reconnect"; data: { event_id: number; reason: string; max_seconds: number } };

async function connectStream(
  client: LumifyClient,
  eventId: number,
  options: StreamScoresOptions
): Promise<Response> {
  const url = client.url(`/v1/events/${eventId}/stream`);

  // Connect-phase timeout only — once headers arrive we clear it so the
  // long-lived SSE body isn't aborted by the client default timeout.
  const connectTimeoutMs =
    options.connectTimeoutMs !== undefined ? options.connectTimeoutMs : client.requestTimeoutMs;
  const connectController = new AbortController();
  let connectTimer: ReturnType<typeof setTimeout> | undefined;
  if (connectTimeoutMs > 0) {
    connectTimer = setTimeout(
      () => connectController.abort(new Error("connect timeout")),
      connectTimeoutMs
    );
  }

  const signals: AbortSignal[] = [connectController.signal];
  if (options.signal) signals.push(options.signal);
  const signal = anySignal(signals);

  let res: Response;
  try {
    res = await client.fetchFn(url, {
      headers: client.authHeaders("text/event-stream"),
      signal,
    });
  } catch (err) {
    throw new ConnectionError(
      `SSE connect to /v1/events/${eventId}/stream failed: ${err instanceof Error ? err.message : String(err)}`,
      { cause: err }
    );
  } finally {
    if (connectTimer) clearTimeout(connectTimer);
  }

  if (!res.ok) {
    let payload: { error?: LumifyErrorPayload } | undefined;
    try {
      payload = await res.json();
    } catch {
      // non-JSON error body — fall through with an undefined payload
    }
    const retryAfterHeader = parseRetryAfter(res.headers.get("retry-after"));
    throw errorFromResponse(res.status, payload?.error, res.headers.get("x-request-id"), {
      retryAfterHeader,
    });
  }

  if (!res.body) {
    throw new ConnectionError(`SSE response for event ${eventId} had no body.`);
  }

  return res;
}

/**
 * Async-iterate live score updates for an event over SSE. Emits only on
 * change (plus a final `done`), so it's far cheaper than polling `get_live_score`.
 *
 * The server caps a single connection at 5 minutes and sends `event: reconnect`
 * just before closing it for that reason (as opposed to `done`, which means the
 * game actually finished). This generator reconnects on that signal
 * automatically and keeps yielding — a long game's stream looks continuous to
 * the caller. The `reconnect` event is still yielded for visibility/telemetry;
 * no action is required in response to it.
 *
 * @example
 * for await (const evt of streamScores(client, eventId)) {
 *   if (evt.event === "score") console.log(evt.data.status, evt.data.clock);
 *   if (evt.event === "done") break;
 * }
 */
export async function* streamScores(
  client: LumifyClient,
  eventId: number,
  options: StreamScoresOptions = {}
): AsyncGenerator<ScoreStreamEvent, void, void> {
  for (;;) {
    const res = await connectStream(client, eventId, options);
    let reconnecting = false;

    for await (const frame of parseSSEStream(res.body!)) {
      if (frame.event === "message" && !frame.data) continue; // keep-alive with no named event
      let parsed: unknown;
      try {
        parsed = frame.data ? JSON.parse(frame.data) : undefined;
      } catch {
        continue; // malformed frame — skip rather than throw mid-stream
      }
      yield { event: frame.event, data: parsed } as ScoreStreamEvent;
      if (frame.event === "reconnect") {
        reconnecting = true;
        break; // tear down this connection below, then loop to reconnect
      }
      if (frame.event === "done" || frame.event === "error") return;
    }

    if (!reconnecting) return; // stream ended without an explicit reconnect signal
  }
}

function parseRetryAfter(header: string | null): number | undefined {
  if (header == null || header === "") return undefined;
  const seconds = Number(header);
  if (!Number.isFinite(seconds) || seconds < 0) return undefined;
  return seconds;
}

function anySignal(signals: AbortSignal[]): AbortSignal {
  if (typeof AbortSignal !== "undefined" && "any" in AbortSignal) {
    return (AbortSignal as unknown as { any(signals: AbortSignal[]): AbortSignal }).any(signals);
  }
  const controller = new AbortController();
  for (const s of signals) {
    if (s.aborted) {
      controller.abort(s.reason);
      break;
    }
    s.addEventListener("abort", () => controller.abort(s.reason), { once: true });
  }
  return controller.signal;
}
